lib/file-source/src/checkpointer.rs

```diff
@@ -94,7 +148,27 @@ impl Checkpointer {
         }
     }
 
+    /// Persist the current checkpoints state to disk, making our best effort to do so in an atomic
+    /// way that allow for recovering the previous state in the event of a crash.
     pub fn write_checkpoints(&mut self) -> Result<usize, io::Error> {
+        // First write the new checkpoints to a tmp file and flush it fully to disk. If vector
+        // dies anywhere during this section, the existing stable file will still be in its current
+        // valid state and we'll be able to recover.
+        let mut f = fs::File::create(&self.tmp_file_path)?;
+        serde_json::to_writer(&mut f, &State::from(&self.checkpoints))?;
+        f.sync_all()?;
+
+        // Once the temp file is fully flushed, rename the tmp file to replace the previous stable
+        // file. This is (almost?) always an atomic operation, which should prevent scenarios where
```

Comment:
According to https://en.wikipedia.org/wiki/Rename_(computing) this is atomic on POSIX but not Windows, but according to https://doc.rust-lang.org/std/fs/fn.rename.html Rust calls the correct equivalent by default. Could you add this to your comment?